contrib/meson.build

```diff
@@ -5,13 +5,21 @@ endif
 
 con2 = configuration_data()
 con2.set('FWUPD_VERSION', fwupd_version)
+con2.set('VERSION', fwupd_version)
```

Comment:
I don't see a need to declare a new `VERSION`, why not just use `FWUPD_VERSION`?

Reply:
So, **great question**.

The basic problem is that we want the SBOM to be accurate (so we want the values substituted with `autoconf`/`automake`, `meson`, `cmake`, `Makefile+sed` etc. rather than hardcoded and then forgotten about on each release. The problem is then when open source projects are used in something like an EDK2 build where the *sources* are used, but the *upstream* build system is ignored. e.g. in EDK2 we might have a build target that uses `local-dxe.c,../gitmodules/libjpeg/jpeg.c` -- e.g. we're not "_building libjpeg and then static linking it in_", we're actually "_building libjpeg_" with a whole parallel build system.

What works now is going something like `uswid --find ~/Code/edk2 --save bom.json` and it searches recursively for any built `bom.json`, `spdx.json` etc files and adds them to a composite BOM to represent the firmware. This works really well if we have a builddir in there that's done the conversion from `bom.json.in` to `bom.json` with the correct substituted values. But it doesn't work if we do the parallel build system thing and the autoconf replacement never gets done.

My proposal in this case would be to "fix up" the found `bom.json.in` files, e.g. `chdir` to the directory, get the _last tagged version_ and the git commit hash of the git submodule and then use those in the composite SBOM. To do that we need to define a set of generic replacement tokens we can use in projects like libssl, e.g. `VERSION` rather than `{project_name}_VERSION`.

I don't know if this idea is insane, or brilliant. Comments most welcome.

Reply:
Implementing this in `uswid` I actually came to the conclusion we don't want to mess about with `.in` files at all. We can just put the `@VERSION@` in the `.json` file and then if we can't autodetect the substitution we can use `NOASSERTION` as the value which makes it a valid file, but would fail validation. Of course, the `@@` keys essentially become ABI, and although I was initially thinking of:

 * `@VERSION@` → `git describe --tags --abbrev=0` e.g. `1.2.3`
 * `@SOURCE_VERSION@` → `git describe --tags` e.g. `1.2.3-250-gfa2371946`

I'm thinking now these should probably be something like

 * `@VCS_TAG@` → `git describe --tags --abbrev=0` e.g. `1.2.3`
 * `@VCS_VERSION@` → `git describe --tags` e.g. `1.2.3-250-gfa2371946`
 * `@VCS_BRANCH@` → `git rev-parse --abbrev-ref HEAD` e.g. `staging`
 * `@VCS_COMMIT@` → `git rev-parse HEAD` e.g. `3090e61ee3452c0478860747de057c0269bfb7b6`

... using `VCS` as a prefix as we probably ought to support `svn`/`hg` as well and `vcs` is what CycloneDX is using.

Also, if we detect two files for the same project (one built, one source -- in the case of actually using the subproject build system) we dedupe them when searching, e.g.

    /home/hughsie/Code/fwupd/venv/build/contrib/bom.json was merged into existing component pkg:github/fwupd/fwupd@2.0.1

Reply:
I would say we should just change `FWUPD_VERSION` to `VERSION`